frontend/controller/actions/group.js

```diff
@@ -521,18 +530,42 @@ export default (sbp('sbp/selectors/register', {
         }
       } catch (e) {
         console.error('gi.actions/group/join failed!', e)
-        sbp('gi.ui/prompt', {
-          heading: L('Failed to join the group'),
-          question: L('Error details:{br_}{err}', { err: e.message, ...LTags() }),
-          primaryButton: L('Close')
-        })
+        throw e
       } finally {
       // If we called join but it didn't result in any actions being sent, we
       // may have left the group. In this case, we execute any pending /remove
       // actions on the contract. This will have no side-effects if /remove on
       // the group contract hasn't been called.
         await sbp('chelonia/contract/release', retainedContracts, { ephemeral: true })
       }
+    }).then(() => {
+      const map: Map<string, any> = sbp('okTurtles.data/get', JOINED_FAILED_KEY)
+      map?.delete(params.contractID)
+    }).catch(e => {
+      let map: Map<string, any> = sbp('okTurtles.data/get', JOINED_FAILED_KEY)
+      if (!map) {
+        map = new Map()
+        sbp('okTurtles.data/set', JOINED_FAILED_KEY, map)
+      }
+
+      map.set(params.contractID, params)
+
+      const scheduleReattempt = () => {
+        if (reattemptTimeoutId === undefined) {
+          reattemptTimeoutId = setTimeout(() => {
+            reattemptTimeoutId = undefined
+            if (map.size === 0) return
+
+            sbp('gi.actions/group/reattemptFailedJoins').catch(e => {
+              console.error('Error running reattemptFailedJoins', e)
+              scheduleReattempt()
+            })
```

Comment:
Two questions:

1. If for some reason the `setTimeout` is forgotten (e.g. browser crashes), will the user be able to re-attempt to join the group by visiting the invite link again?
2. Are you sure the `.catch` here on line 559 will trigger if there's an error? Inside of the `allSettled` there's `return sbp('gi.actions/group/join', params).catch`, and the `.catch` doesn't rethrow... I don't remember the semantics ... will it be considered "not settled" if the inner `.catch` is triggered but doesn't rethrow?

Reply:
Good catch. Originally I'd gone for `Promise.all` but it could settle too quickly. Added missing `then` to convert it to an exception.